### PR TITLE
Assignment: Product Server

### PR DESCRIPTION
**Pull request by alexhcole95**
## Summary

Completed product server assignment with working HTTP methods and tested with Insomnia.

## Changes

1. Completed assignment.
2. Deleted/tidied some files.
3. Moved some directories.

### CODE REVIEW

**Review by jboursiquot — CHANGES_REQUESTED** [review]
Good work. Address the feedback and resubmit.

File: server/main.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	}
+}
+
+func createNewProductHandler(w http.ResponseWriter, r *http.Request) {
+	reqBody, _ := ioutil.ReadAll(r.Body)
+	var item product
+	if err := json.Unmarshal(reqBody, &item); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+	products = append(products, item)
```
[the remark this answers is not in the document]
Comment (alexhcole95) [reply server/main.go:51]: Hey Johnny, I am resubmitting all of these reviews, but unfortunately I am having trouble understanding what is occurring here. When I test this in Insomnia, I can type in the new ID and it applies the ID typed and not 0.

File: server/main.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+		if int64(product.ID) == id {
+			if err = json.NewEncoder(w).Encode(product); err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+			}
+		}
+	}
+}
+
+func createNewProductHandler(w http.ResponseWriter, r *http.Request) {
+	reqBody, _ := ioutil.ReadAll(r.Body)
```
Comment (jboursiquot) [inline server/main.go:46]: You're discarding this error. Handle it.

File: server/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jboursiquot) [reply server/main.go:52]: This means you're submitting the payload with an `id` already set. Yes?

File: server/main.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+func updateProductHandler(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id, err := strconv.ParseInt(vars["id"], 10, 64)
+	if err != nil {
+		http.Error(w, "Unable to convert string to int", http.StatusInternalServerError)
+	}
+
+	var updatedProduct product
+	reqBody, _ := ioutil.ReadAll(r.Body)
```
Comment (jboursiquot) [inline server/main.go:67]: Do not discard this error. Handle it.

File: server/main.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+	defer r.Body.Close()
+	if err = json.Unmarshal(reqBody, &updatedProduct); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+
+	for i, product := range products {
+		if int64(product.ID) == id {
```
Comment (jboursiquot) [inline server/main.go:74]: Once you've found and updated the product, there's no need to keep iterating. Break out of the loop.

**Review by jboursiquot — CHANGES_REQUESTED** [review]

File: server/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexhcole95) [reply server/main.go:52]: That's right, in Insomnia I am typing in the ID when I create a payload.

File: cmd/server/go.mod [hunk 5]
```diff
@@ -0,0 +1,5 @@
+module pivottechschool
```
Comment (jboursiquot) [inline cmd/server/go.mod:1]: Avoid modules in your executable's folder. You only need a single go.mod file at the root of your repo.

File: cmd/server/main.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+	if err := json.NewEncoder(w).Encode(products); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+}
+
+func returnProductHandler(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id, err := strconv.ParseInt(vars["id"], 10, 64)
+	if err != nil {
+		http.Error(w, "Unable to convert string to int", http.StatusInternalServerError)
```
Comment (jboursiquot) [inline cmd/server/main.go:33]: Return a 400 Bad Request to indicate to the caller that the `id` is not valid. This is not an internal server error but a client error.

File: cmd/server/main.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+				w.WriteHeader(http.StatusInternalServerError)
+			}
+		}
+	}
+}
+
+func createNewProductHandler(w http.ResponseWriter, r *http.Request) {
+	reqBody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, "Unable to read request body", http.StatusInternalServerError)
```
Comment (jboursiquot) [inline cmd/server/main.go:48]: Return a 400 Bad Request to indicate to the caller that the id is not valid. This is not an internal server error but a client error.

File: cmd/server/main.go [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+	}
+	defer r.Body.Close()
+
+	var item product
+	if err := json.Unmarshal(reqBody, &item); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+	products = append(products, item)
```
Comment (jboursiquot) [inline cmd/server/main.go:56]: How does your product's ID get set? The client should not be the one to pass it in if that's what's happening.

File: cmd/server/main.go [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+	if err := json.NewEncoder(w).Encode(item); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+}
+
+func updateProductHandler(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id, err := strconv.ParseInt(vars["id"], 10, 64)
+	if err != nil {
+		http.Error(w, "Unable to convert string to int", http.StatusInternalServerError)
```
Comment (jboursiquot) [inline cmd/server/main.go:67]: Return a 400 Bad Request to indicate to the caller that the id is not valid. This is not an internal server error but a client error.

File: cmd/server/main.go [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+			}
+		}
+	}
+}
+
+func deleteProductHandler(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id, err := strconv.ParseInt(vars["id"], 10, 64)
+	if err != nil {
+		http.Error(w, "Unable to convert string to int", http.StatusInternalServerError)
```
Comment (jboursiquot) [inline cmd/server/main.go:99]: Return a 400 Bad Request to indicate to the caller that the id is not valid. This is not an internal server error but a client error.

File: cmd/server/main.go [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+	}
+
+	for _, product := range products {
```
Comment (jboursiquot) [inline cmd/server/main.go:102]: You're not deleting any product here.

**Review by jboursiquot — CHANGES_REQUESTED** [review]

**Review by jboursiquot — APPROVED** [review]

**Comment by jboursiquot** [comment]
You'll need to resolve your conflicts before merging.

**Pull request merged**